### PR TITLE
Run SDK Nightly with CortexXSOAR

### PR DESCRIPTION
**Pull request by jochman**
## Status
- [ ] In Progress
- [x] Ready
- [ ] In Hold - (Reason for hold)

## Related Issues
fixes: https://github.com/demisto/etc/issues/27842

## Description
A few sentences describing the overall goals of the pull request's commits.

## Screenshots
Paste here any images that will help the reviewer

## Minimum version of Demisto
- [ ] 5.0.0
- [ ] 5.5.0
- [ ] 6.0.0

## Does it break backward compatibility?
   - [ ] Yes
       - Further details:
   - [x] No

Passed CircleCI nightly:
https://app.circleci.com/pipelines/github/demisto/content/34407/workflows/a00bfd7c-466b-4219-a4e6-872e5cf38588/jobs/150146

### CODE REVIEW

File: .circleci/config.yml [hunk 1]
```diff
@@ -157,6 +157,104 @@ references:
...
+      command: |
+        if ./Tests/scripts/is_ami.sh ;
+          then
+            ./Tests/scripts/download_demisto_conf.sh
+
+          else
+            ./Tests/lastest_server_build_scripts/download_demisto_conf.sh
+        fi
+  
+  test_download_upload: &test_download_upload
```
Comment (hod-alpert) [inline .circleci/config.yml:173]: This step is used only once, why adding it to references and not the job iteself?

File: .circleci/config.yml [hunk 2]
```diff
@@ -1138,3 +1234,14 @@ workflows:
...
+          context: sdk_nightly_env
+
+  sdk_nightly_trigger:
+    # will initiate when usign the trigger script.
+    when: << pipeline.parameters.demisto_sdk_nightly>>
+    jobs:
+      - Demisto SDK Nightly:
+          context: sdk_nightly_env
+      - Demisto SDK Nightly - Run against CortexXSOAR:
+          context: sdk_nightly_env
```
Comment (hod-alpert) [inline .circleci/config.yml:1247]: Put these jobs in references instead on copy paste them

File: Tests/unlock_integrations.py [hunk 3]
```diff
@@ -0,0 +1,13 @@
+import sys
+
+import demisto_client
+
+from Tests.tools import update_server_configuration
+
+if __name__ == "__main__":
+    try:
+        integration_to_unlock = sys.argv[1]
```
Comment (hod-alpert) [inline Tests/unlock_integrations.py:9]: You're not using this parameter.
remove it or add argparser for it

**Review by hod-alpert — CHANGES_REQUESTED** [review]
Please attach the content test conf PR here too somewhere.

File: .circleci/config.yml (hunk 1 above)
Comment (jochman) [reply .circleci/config.yml:173]: To make the steps much more readable

File: .circleci/config.yml [hunk 4]
```diff
@@ -1138,3 +1234,14 @@ workflows:
     jobs:
       - Demisto SDK Nightly:
           context: sdk_nightly_env
+      - Demisto SDK Nightly - Run against CortexXSOAR:
+          context: sdk_nightly_env
+
+  sdk_nightly_trigger:
+    # will initiate when usign the trigger script.
```
Comment (yaakovi) [inline .circleci/config.yml:1241]:
```suggestion
    # will initiate when using the trigger script.
```

File: .circleci/config.yml [hunk 5]
```diff
@@ -1010,11 +1063,54 @@ jobs:
...
+          name: Wait until server ready
+          shell: /bin/bash
+          when: always
+          command: |
+            python3 ./Tests/scripts/wait_until_server_ready.py "Demisto Marketplace"
+      - run:
+          name: Set Environment Variables
+          when: on_success
+          command: |
+            echo 'export CIRCLE_ARTIFACTS="/home/circleci/project/artifacts"' >> $BASH_ENV
```
Comment (yaakovi) [inline .circleci/config.yml:1088]: you already have it in `prepare_environment`

File: .circleci/config.yml [hunk 6]
```diff
@@ -157,6 +157,104 @@ references:
...
+            ./Tests/lastest_server_build_scripts/download_demisto_conf.sh
+        fi
+  
+  test_download_upload: &test_download_upload
+    run:
+      name: Test Download and Upload commands
+      when: on_success
+      command: |
+        echo "Running upload"
+        demisto-sdk upload -i Packs/HelloWorld/Integrations/ --insecure > artifacts/upload.log
```
Comment (yaakovi) [inline .circleci/config.yml:179]: do this for all logs
```suggestion
        demisto-sdk upload -i Packs/HelloWorld/Integrations/ --insecure > ${CIRCLE_ARTIFACTS}/upload.log
```

File: .circleci/config.yml [hunk 7]
```diff
@@ -157,6 +157,104 @@ references:
...
+        fi
+
+        echo "Running lint"
+        demisto-sdk lint -i Packs/HelloWorld/Integrations/HelloWorld/ > artifacts/lint.log
+        if [[ $? -ne 0 ]]
+        then
+            LINT_CODE=1
+        fi
+        
+        # Check codes
```
Comment (yaakovi) [inline .circleci/config.yml:204]: 1) why aren't you checking the status code for download and upload?
2) currently you ignore the status of format, so why run it?
3) why do all the commands run on the integration and not the pack?

File: .circleci/config.yml (hunk 1 above)
Comment (hod-alpert) [reply .circleci/config.yml:173]: Then make it more readable by separating it to different steps then.

In this single step you're:
1. Uploading.
2. Downloading.
3. Formatting
4. Validating
5. Linting

Each can be a step on its own, which will make the whole process clearer.

And then you could also remove the redundant exit code parsing

**Review by hod-alpert — CHANGES_REQUESTED** [review]
Also, I think you should add a step that checks that all the files are downloaded as expected.
1. That **all** the files have been downloaded
2. That the files have been downloaded as expected (maybe some diff tool or something, I beilve the downloaded files should be the same as the uploaded files, right?)

File: .circleci/config.yml (hunk 5 above)
Comment (jochman) [reply .circleci/config.yml:1088]: didn't work. We worked on it together.

File: .circleci/config.yml (hunk 7 above)
Comment (jochman) [reply .circleci/config.yml:204]: When download or upload are going back, they'll fail the task if not 1. 
I'm ignoring the format outputs as some of its threads will fail on files in TestData.
I'll try to run on the pack. I can't run upload on the pack as it will fail (our server is going up with one helloworld pack).

File: .circleci/config.yml [hunk 8]
```diff
@@ -1010,11 +1063,54 @@ jobs:
...
+          command: |
+            python3 ./Tests/scripts/wait_until_server_ready.py "Demisto Marketplace"
+      - run:
+          name: Set Environment Variables
+          when: on_success
+          command: |
+            echo 'export CIRCLE_ARTIFACTS="/home/circleci/project/artifacts"' >> $BASH_ENV
+            echo "export DEMISTO_API_KEY=$(cat $SECRET_CONF_PATH | jq -r '.temp_apikey')" >> $BASH_ENV
+            echo "export DEMISTO_BASE_URL=\"https://$(cat env_results.json | jq -r '.[0].InstanceDNS')\"" >> $BASH_ENV
+            echo "Server URL: $DEMISTO_BASE_URL"
```
Comment (hod-alpert) [inline .circleci/config.yml:1091]:
```suggestion
            set | grep -E "^DEMISTO_BASE_URL="
```

File: Tests/unlock_integrations.py [hunk 9]
```diff
@@ -0,0 +1,19 @@
...
```
Comment (hod-alpert) [inline Tests/unlock_integrations.py:9]: use argparser

File: .circleci/config.yml (hunk 5 above)
Comment (hod-alpert) [reply .circleci/config.yml:1088]: Just add it [here](https://app.circleci.com/settings/project/github/demisto/content/environment-variables) as variable, it's a constant, no need to add it in run time.

**Review by hod-alpert — CHANGES_REQUESTED** [review]

File: .circleci/config.yml [hunk 10]
```diff
@@ -1099,7 +1052,31 @@ jobs:
...
+          when: on_success
+          command: demisto-sdk download -o Packs/HelloWorld/ -i HelloWorld --insecure -f > ${CIRCLE_ARTIFACTS}/download.log
+      - run:
+          name: Format the integration
+          when: on_success
+          command: |
+            { # try
+            yes | demisto-sdk format -i Packs/HelloWorld/Integrations/HelloWorld -nv -ud > ${CIRCLE_ARTIFACTS}/format.log
+            } || { # catch 
+                echo "Last known exit status: $?"  # DO NOT DELETE THIS LINE. format sometimes will exit with the odd exist statuses.
```
Comment (hod-alpert) [inline .circleci/config.yml:1070]:
```suggestion
                echo "formatting exit status: $?"  # DO NOT DELETE THIS LINE. format sometimes will exit with the odd exist statuses.
```

File: .circleci/config.yml [hunk 11]
```diff
@@ -1010,11 +1017,77 @@ jobs:
...
+          command: pip3 install git+https://github.com/demisto/demisto-sdk.git
+      - run:
+          name: Unlock HelloWorld integration
+          shell: /bin/bash
+          when: on_success
+          command: python3 Tests/unlock_integrations.py HelloWorld
+      - run:
+          name: Upload integration to CortexXSOAR
+          when: on_success
+          command: demisto-sdk upload -i Packs/HelloWorld/Integrations/ --insecure >
```
Comment (hod-alpert) [inline .circleci/config.yml:1058]:
```suggestion
          command: demisto-sdk upload -i Packs/HelloWorld/Integrations/ --insecure 
```

File: .circleci/config.yml [hunk 12]
```diff
@@ -1138,3 +1211,14 @@ workflows:
...
+      - Demisto SDK Nightly - Run against CortexXSOAR:
+          context: sdk_nightly_env
+
+  sdk_nightly_trigger:
+    # will initiate when usign the trigger script.
+    when: << pipeline.parameters.demisto_sdk_nightly>>
+    jobs:
+      - Demisto SDK Nightly:
+          context: sdk_nightly_env
+      - Demisto SDK Nightly - Run against CortexXSOAR:
```
Comment (Itay4) [inline .circleci/config.yml:1223]:
```suggestion
      - Demisto SDK Nightly - Run against Cortex XSOAR:
```

File: .circleci/config.yml [hunk 13]
```diff
@@ -1138,3 +1211,14 @@ workflows:
     jobs:
       - Demisto SDK Nightly:
           context: sdk_nightly_env
+      - Demisto SDK Nightly - Run against CortexXSOAR:
```
Comment (Itay4) [inline .circleci/config.yml:1214]:
```suggestion
      - Demisto SDK Nightly - Run against Cortex XSOAR:
```

File: .circleci/config.yml [hunk 14]
```diff
@@ -1010,11 +1017,77 @@ jobs:
             ./Tests/scripts/slack_notifier.sh 'sdk_faild_steps' ./env_results.json
           when: always
 
+  Demisto SDK Nightly - Run against CortexXSOAR:
```
Comment (Itay4) [inline .circleci/config.yml:1020]:
```suggestion
  Demisto SDK Nightly - Run against Cortex XSOAR:
```

File: .circleci/config.yml [hunk 15]
```diff
@@ -1010,11 +1017,77 @@ jobs:
...
+          shell: /bin/bash
+          when: on_success
+          command: pip3 install git+https://github.com/demisto/demisto-sdk.git
+      - run:
+          name: Unlock HelloWorld integration
+          shell: /bin/bash
+          when: on_success
+          command: python3 Tests/unlock_integrations.py HelloWorld
+      - run:
+          name: Upload integration to CortexXSOAR
```
Comment (Itay4) [inline .circleci/config.yml:1056]:
```suggestion
          name: Upload integration to Cortex XSOAR
```

File: .circleci/config.yml [hunk 16]
```diff
@@ -1010,11 +1017,77 @@ jobs:
...
+          when: on_success
+          command: demisto-sdk upload -i Packs/HelloWorld/Integrations/ --insecure >
+      - run:
+          name: Download integration from CortexXSOAR
```
Comment (Itay4) [inline .circleci/config.yml:1060]:
```suggestion
          name: Download integration from Cortex XSOAR
```

File: .circleci/config.yml [hunk 17]
```diff
@@ -1010,11 +1017,77 @@ jobs:
...
+          when: on_success
+          command: demisto-sdk download -o Packs/HelloWorld/ -i HelloWorld --insecure -f
+      - run:
+          name: Format the integration
+          when: on_success
+          command: |
+            { # try
+            yes | demisto-sdk format -i Packs/HelloWorld/Integrations/HelloWorld -nv -ud
```
Comment (Itay4) [inline .circleci/config.yml:1068]: lets run format and validate on all the pack

File: .circleci/config.yml [hunk 18]
```diff
@@ -1010,11 +1017,77 @@ jobs:
...
+            } || { # catch 
+                echo "Last known exit status: $?"  # DO NOT DELETE THIS LINE. format sometimes will exit with the odd exist statuses.
```
Comment (Itay4) [inline .circleci/config.yml:1070]: lets add TODO here to remove the try-catch with reference to the relevant issue

File: .circleci/config.yml [hunk 19]
```diff
@@ -1010,11 +1017,77 @@ jobs:
...
+            }
+      - run:
+          name: Run validate
+          when: on_success
+          command: demisto-sdk validate -i Packs/HelloWorld/Integrations/HelloWorld/
```
Comment (Itay4) [inline .circleci/config.yml:1075]: on all the pack

File: .circleci/config.yml [hunk 20]
```diff
@@ -1010,11 +1017,77 @@ jobs:
...
+      - run:
+          name: Run lint
+          when: on_success
+          command: demisto-sdk lint -i Packs/HelloWorld/Integrations/HelloWorld/
```
Comment (Itay4) [inline .circleci/config.yml:1079]: on all the pack

File: Pipfile [hunk 21]
```diff
@@ -4,6 +4,7 @@ url = "https://pypi.org/simple"
 verify_ssl = true
 
 [dev-packages]
+pylint = "*"
```
Comment (Itay4) [inline Pipfile:7]: why is this needed?

**Comment by Itay4** [comment]
we have too much going on in `Tests`, lets move the scripts to be in a `Tests/sdk_nightly` dir

File: Tests/create_entities_for_nightly_sdk.py [hunk 22]
```diff
@@ -0,0 +1,135 @@
...
+        'id': name,
+        'layout': layout_name
+    })
+    incident_path = incident_type_path / f'incidenttype-{name.replace(" ", "_")}.json'
+    with open(incident_path, 'w+') as stream:
+        json.dump(incident_type, stream, indent=4)
+    return str(incident_path)
+
+
+def upload_to_sdk(path: Union[Path, str], *args: str):
```
Comment (yaakovi) [inline Tests/create_entities_for_nightly_sdk.py:98]: why is this part of the script and not run after it?

File: Tests/create_entities_for_nightly_sdk.py [hunk 23]
```diff
@@ -0,0 +1,135 @@
...
+        print(f'{arg} was uploaded to Cortex XSOAR')
+
+
+def main():
+    parser = argparse.ArgumentParser(description="Creates incident field, incident type and a layout in a given pack.")
+    parser.add_argument('pack_name')
+    args = parser.parse_args()
+    pack_path = Path('Packs') / args.pack_name
+    layout_name = 'Hello World Test Layout'
+    uploaded_entities = [
```
Comment (yaakovi) [inline Tests/create_entities_for_nightly_sdk.py:124]: save them to build artifacts as well

**Review by yaakovi — APPROVED** [review]

File: Tests/create_entities_for_nightly_sdk.py [hunk 24]
```diff
@@ -0,0 +1,135 @@
...
+        raise AttributeError(f'Could not find output of the command. \nstdout={stdout}\nstderr={stderr}')
+    trimmed = searched_re.group()
+    for arg in args:
+        assert arg.split('/')[-1] in trimmed, f'Could not upload {arg}.\nstdout={stdout}\nstderr={stderr}'
+        print(f'{arg} was uploaded to Cortex XSOAR')
+
+
+def main():
+    parser = argparse.ArgumentParser(description="Creates incident field, incident type and a layout in a given pack.")
+    parser.add_argument('pack_name')
```
Comment (yaakovi) [inline Tests/create_entities_for_nightly_sdk.py:120]: you get a `pack_name` and use `Hello World` everywhere

File: Tests/create_entities_for_nightly_sdk.py (hunk 22 above)
Comment (jochman) [reply Tests/create_entities_for_nightly_sdk.py:98]: Bug I tried to workaround.
Trying without it

**Comment by Itay4** [comment]
why do we need the `__init__.py` file ?

File: .circleci/config.yml [hunk 25]
```diff
@@ -337,6 +385,25 @@ references:
...
               SERVER_URL=$(jq -r 'select(.[].Role == "Demisto Marketplace") | .[].InstanceDNS' ./env_results.json)
               python3 ./Utils/comment_on_pr.py -p $PULL_REQUEST_NUMBER -c "Instance is ready. Server link: https://$SERVER_URL, Build link: $CIRCLE_BUILD_URL"
+
+  upload_entities_to_cortex_xsoar: &upload_entities_to_cortex_xsoar
+    run:
+        name: Upload entities to Cortex XSOAR
+        command: |
+          demisto-sdk upload -i Packs/HelloWorld/Integrations/ --insecure
+          demisto-sdk upload -i Packs/HelloWorld/TestPlaybooks/playbook-HelloWorld-Test.yml --insecure
+          # TODO: After fixing the demisto-sdk upload, upload from here instead of Tests/sdknightly/create_entities_for_nightly_sdk.py
```
Comment (Itay4) [inline .circleci/config.yml:395]: is there issue for this?

File: .circleci/config.yml [hunk 26]
```diff
@@ -1010,11 +1034,79 @@ jobs:
...
+          command: |
+            demisto-sdk download -o Packs/HelloWorld/ -i HelloWorld --insecure -f
+            demisto-sdk download -o Packs/HelloWorld/ -i HelloWorld-Test --insecure -f
+            demisto-sdk download -o Packs/HelloWorld/ -i "Hello World Incident Test" --insecure -f
+            demisto-sdk download -o Packs/HelloWorld/ -i "Hello World Alert Test" --insecure -f
+      - run:
+          name: Format the entities
+          command: |
+            { # try
+            yes | demisto-sdk format -i Packs/HelloWorld/ -nv -ud
```
Comment (hod-alpert) [inline .circleci/config.yml:1081]:
```suggestion
            demisto-sdk format -i Packs/HelloWorld/ -nv -ud -y
```

File: .circleci/config.yml [hunk 27]
```diff
@@ -1010,11 +1034,79 @@ jobs:
...
+      - run:
+          name: Download entities from Cortex XSOAR
+          command: |
+            demisto-sdk download -o Packs/HelloWorld/ -i HelloWorld --insecure -f
+            demisto-sdk download -o Packs/HelloWorld/ -i HelloWorld-Test --insecure -f
+            demisto-sdk download -o Packs/HelloWorld/ -i "Hello World Incident Test" --insecure -f
+            demisto-sdk download -o Packs/HelloWorld/ -i "Hello World Alert Test" --insecure -f
+      - run:
+          name: Format the entities
+          command: |
```
Comment (hod-alpert) [inline .circleci/config.yml:1079]: Add a `shell: /bin/bash` line to the command and then you can remove the try and catch.

When specifying `shell: /bin/bash` it will return the last command exit code and will not fail at first failure like it behaves without it.

The command should look like this:
```
- run:
          name: Format the entities
          shell: /bin/bash
          command: |
             demisto-sdk format -i Packs/HelloWorld/ -nv -ud -y
             echo "format exit status: $?" 
```

Comment (hod-alpert) [reply .circleci/config.yml:1079]: What if the format fails for the right reasons?
Shouldn't you add some logic about that to see that you fail the build in such cases?

Comment (jochman) [reply .circleci/config.yml:1079]: there's an open bug. after the bug will be fixed we can remove it.

**Comment by jochman** [comment]
> why do we need the `__init__.py` file ?

we don't. Auto created by pycharm

File: .circleci/config.yml (hunk 25 above)
Comment (jochman) [reply .circleci/config.yml:395]: Yeah, I opened one and @ShahafBenYakir said there isn't one. reopening https://github.com/demisto/etc/issues/29578

File: .circleci/config.yml (hunk 27 above)
Comment (jochman) [reply .circleci/config.yml:1080]: Issue https://github.com/demisto/etc/issues/29575

**Review by hod-alpert — APPROVED** [review]

**Review by Itay4 — APPROVED** [review]

File: .circleci/config.yml (hunk 25 above)
Comment (ShahafBenYakir) [reply .circleci/config.yml]: This is hot fake news is spread...
